pinot-core/src/main/java/org/apache/pinot/core/segment/processing/mapper/SegmentMapper.java

```diff
@@ -141,28 +146,43 @@ private Map<String, GenericRowFileManager> doMap()
               RecordReaderFactory.getRecordReader(recordReaderFileConfig._fileFormat, recordReaderFileConfig._dataFile,
                   recordReaderFileConfig._fieldsToRead, recordReaderFileConfig._recordReaderConfig);
           mapAndTransformRow(recordReader, reuse, observer, count, totalCount);
+          _recordReaderFileConfigs.get(i)._recordReader = recordReader;
         } finally {
-          if (recordReader != null) {
+          if (recordReader != null && !recordReader.hasNext()) {
             recordReader.close();
           }
         }
       } else {
+        if (!recordReader.hasNext()) {
+          LOGGER.info("Skipping record reader as it is already processed at index: {}", i);
+          count++;
+          continue;
+        }
         mapAndTransformRow(recordReader, reuse, observer, count, totalCount);
+        _recordReaderFileConfigs.get(i)._recordReader = recordReader;
+      }
+      if (!_adaptiveSizeBasedWriter.canWrite()) {
```

Comment:
Instead of checking for this again,  we can have mapAndTransformRow pass a return value to terminate this loop ?

Reply:
makes sense. I have added this logic.

Reply:
On further inspection I have kept the check only in one place. Right after `mapAndTransformRow`. If we got kicked out of `mapAndTransformRow` because of size constraints this check should be enough. 

Please let me know if returning a boolean value makes more sense. I will change accordingly.

Reply:
Could you check if you can keep this constraints logic where its needed (in mapAndTransformRow) and add the corresponding logs there itself. The method can return boolean to break this loop?

In the log.info, please how many readers we completed, how many pending. Or you can add this log in SegmentProcessorFramework where you have the overall counts?

Reply:
For the first part I have moved the logic.

We are already logging the number of record readers already processed, total count, number of bytes written when we break off etc. in `mapAndTransformRow`. It looks something like this:
<img width="706" alt="Screenshot 2024-01-11 at 11 40 32 PM" src="https://github.com/apache/pinot/assets/15700987/801ef796-8a2d-4d30-84a3-01a720063a19">
Do we need any more?